File: src/reframe/catalog/selected_portion.py
"""Fail-closed delivery of independent, completed input segments."""
from functools import lru_cache
import json
from pathlib import Path

from src.reframe.spoiler.analysis_boundary import SEPARATED_ANALYSIS_EDITIONS, current_position

DATA_PATH = Path(__file__).resolve().parents[3]/'data/production/selected_portion_analysis.json'


def validate_package(package):
    if package.get('version') != 'selected-portion-v1-20260909':
        raise ValueError('Unsupported selected-portion publication')
    films = package['films']
    if {(f['work_id'],f['edition_id']) for f in films} != SEPARATED_ANALYSIS_EDITIONS or len(films) != 3:
        raise ValueError('Unexpected selected-portion editions')
    for film in films:
        runtime = film['runtime_ms']
        if type(runtime) is not int or runtime <= 0:
            raise ValueError('Invalid edition runtime')
        seen = set()
        previous = -1
        for segment in film['segments']:
            bounds = [segment[k] for k in ('start_ms','input_end_ms','available_after_ms')]
            if any(type(v) is not int for v in bounds) or not 0 <= bounds[0] < bounds[1] <= bounds[2] <= runtime:
                raise ValueError('Invalid segment input boundary')
            if bounds[0] < previous or segment['segment_id'] in seen:
                raise ValueError('Duplicate or overlapping segments')
            previous = bounds[1]
            seen.add(segment['segment_id'])
            if segment['source_scope'] != 'INDEPENDENT_SEGMENT_ONLY':
                raise ValueError('Full-film synthesis is not a watched-portion source')
            if len(segment['source_response_sha256']) != 64:
                raise ValueError('Missing immutable observation provenance')
            if not segment['summary'] or not segment['observations']:
                raise ValueError('Missing segment observations')
            for observation in segment['observations']:
                for key in ('timestamp_ms','evidence_end_ms'):
                    v = observation[key]
                    if type(v) is not int or not bounds[0] <= v <= bounds[1]:
                        raise ValueError('Observation exceeds its source input')
    return package


@lru_cache(maxsize=1)
def load_package():
    return validate_package(json.loads(DATA_PATH.read_text(encoding='utf-8')))


def selected_portion(work_id, edition_id, viewer, selected_ms=None):
    if (work_id, edition_id) not in SEPARATED_ANALYSIS_EDITIONS:
        return None
    film = next(f for f in load_package()['films'] if (f['work_id'],f['edition_id']) == (work_id,edition_id))
    return selected_portion_from_film(film, viewer, selected_ms)


def selected_portion_from_film(film, viewer, selected_ms=None):
    """Shape already validated observations with the same per-edition boundary."""
    work_id, edition_id = film['work_id'], film['edition_id']
    saved = min(film['runtime_ms'], current_position(viewer,work_id,edition_id))
    if selected_ms is not None and (type(selected_ms) is not int or selected_ms < 0):
        raise ValueError('Invalid selected position')
    position = saved if selected_ms is None else min(saved,selected_ms)
    # Filter BEFORE constructing a DTO. No future summary, title, ID, frame,
    # count, or text is delivered and then merely hidden by the browser.
    completed = [s for s in film['segments'] if 0 < s['available_after_ms'] <= position]
    return {
        'work_id':work_id, 'edition_id':edition_id, 'selected_progress_ms':position,
        'covered_until_ms':max((s['input_end_ms'] for s in completed),default=0),
        'segment_interval_ms':film['segment_interval_ms'], 'mode':'SELECTED_PORTION',
        'human_review_status':'NOT_REVIEWED', 'model_id':film['model_id'],
        'retrospective_available':position >= film['runtime_ms'],
        'segments':[{
            'segment_id':s['segment_id'], 'start_ms':s['start_ms'], 'end_ms':s['input_end_ms'],
            'available_after_ms':s['available_after_ms'], 'summary':s['summary'],
            'reading':s['reading'], 'observations':[dict(o) for o in s['observations']],
            'source_method':s['source_method'], 'source_response_sha256':s['source_response_sha256'],
        } for s in completed],
    }

File: apps/web/src/services/selectedPortion.ts
export function supportsSelectedPortion(movie: string, edition: string): boolean {
  return [
    'the-bat-whispers-1930:tbw-fullscreen-archive',
    'the-greene-murder-case-1929:gmc-archive-1929',
    'the-thirteenth-chair-1929:ttc-archive-1929',
  ].includes(`${movie}:${edition}`);
}

export interface PortionSegment {
  segment_id: string;
  start_ms: number;
  end_ms: number;
  available_after_ms: number;
  summary: string;
  reading: string;
  source_method: 'SAMPLED_STILLS_ONLY' | 'AUDIO_AND_SAMPLED_STILLS';
  observations: { timestamp_ms: number; evidence_end_ms: number; text: string; uncertainty: string }[];
}

export interface SelectedPortion {
  work_id: string;
  edition_id: string;
  selected_progress_ms: number;
  covered_until_ms: number;
  mode: 'SELECTED_PORTION';
  human_review_status: 'NOT_REVIEWED';
  model_id: string;
  segments: PortionSegment[];
}

export function validateSelectedPortion(data: SelectedPortion, movie: string, edition: string, position: number): SelectedPortion {
  const time = (v: number) => Number.isSafeInteger(v) && v >= 0;
  if (data?.work_id !== movie || data.edition_id !== edition || data.mode !== 'SELECTED_PORTION' ||
      !time(data.selected_progress_ms) || data.selected_progress_ms > position ||
      !time(data.covered_until_ms) || data.covered_until_ms > data.selected_progress_ms || !Array.isArray(data.segments)) {
    throw new Error('The selected-portion response did not match this viewing position.');
  }
  for (const segment of data.segments) {
    if (![segment.start_ms, segment.end_ms, segment.available_after_ms].every(time) ||
        segment.start_ms >= segment.end_ms || segment.end_ms > segment.available_after_ms ||
        segment.available_after_ms > data.selected_progress_ms ||
        typeof segment.summary !== 'string' || typeof segment.reading !== 'string' || !Array.isArray(segment.observations) ||
        segment.observations.some(o => !time(o.timestamp_ms) || !time(o.evidence_end_ms) ||
          o.timestamp_ms < segment.start_ms || o.timestamp_ms > segment.end_ms ||
          o.evidence_end_ms < segment.start_ms || o.evidence_end_ms > segment.end_ms || typeof o.text !== 'string')) {
      throw new Error('A segment exceeded the selected viewing position.');
    }
  }
  return data;
}
